File: ru.kmorozov.gbd.db/src/main/java/ru/kmorozov/db/utils/gson/IPageAdapter.java
package ru.kmorozov.db.utils.gson;

import com.google.gson.*;
import ru.kmorozov.db.core.logic.model.book.google.GooglePageInfo;
import ru.kmorozov.gbd.core.logic.model.book.base.IPage;
import ru.kmorozov.db.core.logic.model.book.shpl.ShplPage;

import java.lang.reflect.Type;

/**
 * Created by sbt-morozov-kv on 17.11.2016.
 */
public class IPageAdapter implements JsonSerializer<IPage>, JsonDeserializer<IPage> {

    @Override
    public IPage deserialize(final JsonElement json, final Type typeOfT, final JsonDeserializationContext context) throws JsonParseException {
        final JsonObject jsonObject = json.getAsJsonObject();

        return context.deserialize(jsonObject, null == jsonObject.get("pid") ? ShplPage.class : GooglePageInfo.class);
    }

    @Override
    public JsonElement serialize(final IPage src, final Type typeOfSrc, final JsonSerializationContext context) {
        return context.serialize(src);
    }
}

File: ru.kmorozov.gbd.db/src/main/java/ru/kmorozov/db/utils/gson/IPageAdapter.kt
package ru.kmorozov.db.utils.gson

import com.google.gson.*
import ru.kmorozov.db.core.logic.model.book.google.GooglePageInfo
import ru.kmorozov.gbd.core.logic.model.book.base.IPage
import ru.kmorozov.db.core.logic.model.book.shpl.ShplPage

import java.lang.reflect.Type

/**
 * Created by sbt-morozov-kv on 17.11.2016.
 */
class IPageAdapter : JsonSerializer<IPage>, JsonDeserializer<IPage> {

    @Throws(JsonParseException::class)
    override fun deserialize(json: JsonElement, typeOfT: Type, context: JsonDeserializationContext): IPage {
        val jsonObject = json.asJsonObject

        return context.deserialize(jsonObject, if (null == jsonObject.get("pid")) ShplPage::class.java else GooglePageInfo::class.java)
    }

    override fun serialize(src: IPage, typeOfSrc: Type, context: JsonSerializationContext): JsonElement {
        return context.serialize(src)
    }
}
